File: src/utils/salary-calculator.ts
/**
 * Malta Salary Calculator
 * Excel formüllerine uygun hesaplama mantığı
 * Kaynak: Payroll Working.xlsx ve .agent/rules/malta-payroll-rules.md
 */

import {
    getTaxBracketsForYear,
    getSSCRatesForYear,
    getCOLAForMonth,
    isBornBefore1962,
    type TaxBracket,
    type SSCRates,
    type SSCCategory,
    type TaxRateType,
} from "@/config/malta-tax-config";

import {
    Month,
    weeksPerMonth,
    type MonthlySalaryInput,
    type MonthlySalaryOutput,
    type SalaryCalculatorConfig,
} from "@/types/salary-calculator-type";

// ==================== SSC HESAPLAMA ====================

/**
 * Haftalık SSC Base hesaplar (cap uygulanır)
 * Excel formülü: IF((B*12)/52 < CAP, (B*12)/52 * weeks, CAP * weeks)
 */
function calculateSSCBase(
    basicSalary: number,
    weeksInMonth: number,
    weeklyCap: number
): number {
    const weeklyEquivalent = (basicSalary * 12) / 52;

    if (weeklyEquivalent < weeklyCap) {
        return weeklyEquivalent * weeksInMonth;
    }
    return weeklyCap * weeksInMonth;
}

/**
 * SSC Tax hesaplar (kategori bazlı)
 * Excel formülleri: Kategori A, B, C, D için farklı hesaplamalar
 */
function calculateSSCTax(
    sscBase: number,
    weeksInMonth: number,
    sscCategory: SSCCategory,
    sscRates: SSCRates,
    isBornBefore1962Flag: boolean
): number {
    switch (sscCategory) {
        case 'A':
            // 18 yaş altı - sabit haftalık oran
            return sscRates.categoryA * weeksInMonth;

        case 'B':
            // Part-time veya minimum ücret altı
            // MIN(categoryB * weeks, sscBase * 0.10)
            return Math.min(
                sscRates.categoryB * weeksInMonth,
                Math.round(sscBase * 0.10 * 100) / 100
            );

        case 'C':
            // Tam zamanlı çalışan
            if (isBornBefore1962Flag) {
                return Math.min(
                    sscRates.categoryCOld * weeksInMonth,
                    Math.round(sscBase * 0.10 * 100) / 100
                );
            }
            return Math.min(
                sscRates.categoryCNew * weeksInMonth,
                Math.round(sscBase * 0.10 * 100) / 100
            );

        case 'D':
            // Tavan üstü gelir - sabit haftalık oran
            if (isBornBefore1962Flag) {
                return sscRates.categoryDOld * weeksInMonth;
            }
            return sscRates.categoryDNew * weeksInMonth;

        default:
            return 0;
    }
}

// ==================== VERGİ HESAPLAMA ====================

/**
 * Yıllık gelir için uygulanacak vergi dilimini bulur
 */
function findTaxBracket(annualIncome: number, brackets: TaxBracket[]): TaxBracket | null {
    for (const bracket of brackets) {
        if (annualIncome >= bracket.min && annualIncome <= bracket.max) {
            return bracket;
        }
    }
    // Gelir son dilimin üstündeyse son dilimi kullan
    return brackets[brackets.length - 1];
}

/**
 * Aylık gelir vergisi hesaplar (kümülatif sistem)
 * 
 * İlk ay: (((incomeBase * 12) * rate) - deduction) / 12
 * Sonraki aylar: (((cumulativeIncomeBase / month * 12) * rate) - deduction) / 12 * month - previousCumulativeTax
 */
function calculateMonthlyIncomeTax(
    incomeBase: number,
    cumulativeIncomeBase: number,
    previousCumulativeTax: number,
    monthIndex: number,
    annualGross: number,
    taxBrackets: TaxBracket[]
): number {
    const bracket = findTaxBracket(annualGross, taxBrackets);

    if (!bracket || bracket.rate === 0) {
        return 0;
    }

    if (monthIndex === 0) {
        // İlk ay
        const yearlyProjectedTax = (incomeBase * 12 * bracket.rate) - bracket.deduction;
        return Math.max(0, yearlyProjectedTax / 12);
    } else {
        // Sonraki aylar (kümülatif hesaplama)
        const monthNumber = monthIndex + 1;
        const projectedAnnualIncome = (cumulativeIncomeBase / monthNumber) * 12;
        const yearlyProjectedTax = (projectedAnnualIncome * bracket.rate) - bracket.deduction;
        const monthlyTax = (yearlyProjectedTax / 12) * monthNumber - previousCumulativeTax;
        return Math.max(0, monthlyTax);
    }
}

// ==================== ANA HESAPLAMA FONKSİYONU ====================

/**
 * Varsayılan konfigurasyon
 */
export const defaultConfig: SalaryCalculatorConfig = {
    year: 2025,
    taxRateType: 'single',
    sscCategory: 'C',
    birthDate: new Date(1990, 0, 1),
    yearlyNonTaxBenefit: 1170,  // €97.50 * 12
    yearlyTaxableBenefit: 1170, // €97.50 * 12
    enableCOLA: true,           // COLA otomatik hesaplanır
};

/**
 * Aylık maaş kesintilerini hesaplar
 */
export function calculateMonthlyDeductions(
    salaryInput: MonthlySalaryInput[],
    config: SalaryCalculatorConfig = defaultConfig
): MonthlySalaryOutput[] {
    // Config'den değerleri al
    const taxBrackets = getTaxBracketsForYear(config.year, config.taxRateType);
    const sscRates = getSSCRatesForYear(config.year);
    const isBornBefore1962Flag = isBornBefore1962(config.birthDate);

    // Haftalık SSC cap (doğum tarihine göre)
    const weeklyCap = isBornBefore1962Flag
        ? sscRates.weeklyCapOld
        : sscRates.weeklyCapNew;

    // Aylık sabit değerler
    const monthlyNonTaxBenefit = config.yearlyNonTaxBenefit / 12;
    const monthlyTaxBenefit = config.yearlyTaxableBenefit / 12;

    // Yıllık brüt (vergi dilimi belirlemek için)
    const totalAnnualGross = salaryInput.reduce((sum, s) => sum + s.grossWage, 0);

    // Kümülatif değerler
    let cumulativeIncomeBase = 0;
    let cumulativeTax = 0;

    return salaryInput.map((input, index) => {
        const { month } = input;
        // NaN koruması: grossWage, bonus, governmentBonus ve allowanceBonus için
        const grossWage = isNaN(Number(input.grossWage)) ? 0 : Number(input.grossWage);
        const inputBonus = isNaN(Number(input.bonus)) ? 0 : Number(input.bonus);
        const inputGovBonus = isNaN(Number(input.governmentBonus)) ? 0 : Number(input.governmentBonus);
        const safeAllowanceBonus = isNaN(Number(input.allowanceBonus)) ? 0 : Number(input.allowanceBonus);

        // Hafta sayısı: override varsa onu kullan, yoksa input'tan, o da yoksa default
        const weeksInMonth = config.weeksPerMonthOverride ?? input.weeksInMonth ?? weeksPerMonth[month];

        // Temel hesaplamalar
        const basicSalary = grossWage;
        const nonTaxBenefit = monthlyNonTaxBenefit;
        const taxBenefit = monthlyTaxBenefit;

        // Bonus ve Government Bonus (COLA)
        const bonus = inputBonus || (config.monthlyBonus ?? 0);
        // COLA: Eğer enableCOLA true ise otomatik olarak ay bazlı COLA ekle
        const autoCOLA = config.enableCOLA ? getCOLAForMonth(config.year, month) : 0;
        const governmentBonus = inputGovBonus || autoCOLA;
        const halfAdditional = safeAllowanceBonus / 2;

        // Brüt toplam
        const grossTotal = basicSalary + nonTaxBenefit + taxBenefit + bonus + governmentBonus + halfAdditional;

        // SSC hesaplama
        const sscBase = calculateSSCBase(basicSalary, weeksInMonth, weeklyCap);
        const sscTax = calculateSSCTax(
            sscBase,
            weeksInMonth,
            config.sscCategory,
            sscRates,
            isBornBefore1962Flag
        );

        // Income Base (vergiye tabi gelir)
        // Excel: Income Base = Gross Total - Non-Tax Benefit
        const incomeBase = grossTotal - nonTaxBenefit;

        // Kümülatif gelir tabanı
        cumulativeIncomeBase += incomeBase;

        // Gelir vergisi hesaplama (kümülatif sistem)
        const incomeTax = calculateMonthlyIncomeTax(
            incomeBase,
            cumulativeIncomeBase,
            cumulativeTax,
            index,
            totalAnnualGross,
            taxBrackets
        );

        // Kümülatif vergi güncelle
        cumulativeTax += incomeTax;

        // Net maaş
        // Excel: Net = Gross Total - SSC Tax - Income Tax
        const net = grossTotal - sscTax - incomeTax;
        const paid = net + halfAdditional; // Allowance bonus'un diğer yarısı
        const discr = grossTotal - paid;

        return {
            month,
            grossWage,
            basicSalary,
            nonTaxBenefit: Math.round(nonTaxBenefit * 100) / 100,
            taxBenefit: Math.round(taxBenefit * 100) / 100,
            bonus,
            governmentBonus,
            grossTotal: Math.round(grossTotal * 100) / 100,
            sscBase: Math.round(sscBase * 100) / 100,
            sscTax: Math.round(sscTax * 100) / 100,
            incomeBase: Math.round(incomeBase * 100) / 100,
            cumulativeIncomeBase: Math.round(cumulativeIncomeBase * 100) / 100,
            cumulativeTax: Math.round(cumulativeTax * 100) / 100,
            incomeTax: Math.round(incomeTax * 100) / 100,
            net: Math.round(net * 100) / 100,
            paid: Math.round(paid * 100) / 100,
            discr: Math.round(discr * 100) / 100,
        };
    });
}

// ==================== ESKİ FONKSİYON (GERİYE UYUMLULUK) ====================

/**
 * @deprecated Yeni calculateMonthlyDeductions fonksiyonunu config ile kullanın
 */
export function calculateMonthlyDeductionsLegacy(
    salaryInput: MonthlySalaryInput[]
): MonthlySalaryOutput[] {
    return calculateMonthlyDeductions(salaryInput, defaultConfig);
}

File: ios-app/Packages/CalculationKit/Sources/CalculationKit/Salary/SalaryCalculator.swift
import Foundation

/// Malta Salary Calculator — cumulative income tax, SSC, and COLA.
///
/// Ports the TypeScript implementation in `src/utils/salary-calculator.ts`
/// to Swift using `Decimal`-only arithmetic for financial accuracy.
///
/// Reference: `src/utils/salary-calculator.ts`
public struct SalaryCalculator: Sendable {
    /// The salary calculator configuration (tax type, SSC category, benefits, etc.).
    public let config: SalaryCalculatorConfig
    /// The Malta tax configuration containing brackets, SSC rates, and COLA.
    public let taxConfig: MaltaTaxConfig

    /// Creates a new salary calculator with the given configuration.
    public init(config: SalaryCalculatorConfig, taxConfig: MaltaTaxConfig) {
        self.config = config
        self.taxConfig = taxConfig
    }

    /// Computes monthly salary breakdowns for each input period.
    ///
    /// - Parameter inputs: An array of monthly salary inputs (typically 12 months).
    /// - Returns: An array of `SalaryOutput` values, one per input month.
    /// - Throws: `CalculationError.invalidYear` if the year is not in config,
    ///           `CalculationError.corruptedConfig` if brackets are missing.
    public func calculate(inputs: [SalaryInput]) throws -> [SalaryOutput] {
        guard let yearConfig = taxConfig.years[config.year] else {
            throw CalculationError.invalidYear(config.year)
        }

        let effectiveTaxType = resolveTaxRateType(
            year: config.year,
            simpleType: config.simpleTaxType,
            childCount: config.childCount
        )

        guard let taxBrackets = yearConfig.brackets[effectiveTaxType]
            ?? yearConfig.brackets[.single]
        else {
            throw CalculationError.corruptedConfig(
                reason: "No brackets for \(effectiveTaxType) in \(config.year)"
            )
        }

        let sscCalculator = SSCCalculator(
            sscRates: yearConfig.ssc,
            sscCategory: config.sscCategory,
            isBornBefore1962: DateHelpers.isBornBefore1962(config.birthDate)
        )

        let monthlyNonTaxBenefit = config.yearlyNonTaxBenefit / 12
        let monthlyTaxBenefit = config.yearlyTaxableBenefit / 12
        let totalAnnualGross = inputs.map(\.grossWage).reduce(0, +)

        var cumulativeIncomeBase: Money = 0
        var cumulativeTax: Money = 0
        var outputs: [SalaryOutput] = []
        outputs.reserveCapacity(inputs.count)

        for (index, input) in inputs.enumerated() {
            let grossWage = input.grossWage.nonNegative
            let inputBonus = input.bonus.nonNegative
            let inputGovBonus = input.governmentBonus.nonNegative
            let safeAllowanceBonus = input.allowanceBonus.nonNegative

            let weeksInMonth = config.weeksPerMonthOverride
                ?? input.weeksInMonth
                ?? DateHelpers.weeksForMonth(year: config.year, month: input.month)

            let basicSalary = grossWage
            let nonTaxBenefit = monthlyNonTaxBenefit
            let taxBenefit = monthlyTaxBenefit

            let monthlyBonusFallback = config.monthlyBonus
            let bonus = inputBonus > 0 ? inputBonus : monthlyBonusFallback
            let autoCOLA: Money = config.enableCOLA
                ? yearConfig.cola.amount(for: input.month)
                : 0
            let governmentBonus = inputGovBonus > 0 ? inputGovBonus : autoCOLA
            let halfAdditional = safeAllowanceBonus / 2

            let grossTotal = basicSalary + nonTaxBenefit + taxBenefit
                + bonus + governmentBonus + halfAdditional

            let sscBase = sscCalculator.calculateBase(
                basicSalary: basicSalary,
                weeksInMonth: weeksInMonth
            )
            let sscTax = sscCalculator.calculateTax(
                sscBase: sscBase,
                weeksInMonth: weeksInMonth
            )

            let incomeBase = grossTotal - nonTaxBenefit
            cumulativeIncomeBase += incomeBase

            let incomeTax = calculateMonthlyIncomeTax(
                incomeBase: incomeBase,
                cumulativeIncomeBase: cumulativeIncomeBase,
                previousCumulativeTax: cumulativeTax,
                monthIndex: index,
                annualGross: totalAnnualGross,
                taxBrackets: taxBrackets
            )
            cumulativeTax += incomeTax

            let net = grossTotal - sscTax - incomeTax
            let paid = net + halfAdditional
            let discr = grossTotal - paid

            outputs.append(SalaryOutput(
                month: input.month,
                grossWage: grossWage,
                basicSalary: basicSalary,
                nonTaxBenefit: nonTaxBenefit.rounded(),
                taxBenefit: taxBenefit.rounded(),
                bonus: bonus,
                governmentBonus: governmentBonus,
                grossTotal: grossTotal.rounded(),
                sscBase: sscBase.rounded(),
                sscTax: sscTax.rounded(),
                incomeBase: incomeBase.rounded(),
                cumulativeIncomeBase: cumulativeIncomeBase.rounded(),
                cumulativeTax: cumulativeTax.rounded(),
                incomeTax: incomeTax.rounded(),
                net: net.rounded(),
                paid: paid.rounded(),
                discr: discr.rounded()
            ))
        }

        return outputs
    }

    /// Monthly income tax using Malta's cumulative system.
    ///
    /// Projects annual income from the year-to-date cumulative base,
    /// applies the matching bracket rate and deduction, then subtracts
    /// previously paid cumulative tax to derive this month's portion.
    private func calculateMonthlyIncomeTax(
        incomeBase: Money,
        cumulativeIncomeBase: Money,
        previousCumulativeTax: Money,
        monthIndex: Int,
        annualGross: Money,
        taxBrackets: [MaltaTaxConfig.TaxBracket]
    ) -> Money {
        guard let bracket = findTaxBracket(for: annualGross, in: taxBrackets),
              bracket.rate > 0
        else { return 0 }

        if monthIndex == 0 {
            let yearlyProjectedTax = incomeBase * 12 * bracket.rate - bracket.deduction
            return Swift.max(0, yearlyProjectedTax / 12)
        } else {
            let monthNumber = Decimal(monthIndex + 1)
            let projectedAnnualIncome = (cumulativeIncomeBase / monthNumber) * 12
            let yearlyProjectedTax = projectedAnnualIncome * bracket.rate - bracket.deduction
            let monthlyTax = (yearlyProjectedTax / 12) * monthNumber - previousCumulativeTax
            return Swift.max(0, monthlyTax)
        }
    }

    /// Finds the tax bracket that contains the given annual income.
    ///
    /// Falls back to the last bracket if no exact match is found.
    private func findTaxBracket(
        for annualIncome: Money,
        in brackets: [MaltaTaxConfig.TaxBracket]
    ) -> MaltaTaxConfig.TaxBracket? {
        for bracket in brackets where annualIncome >= bracket.min && annualIncome <= bracket.max {
            return bracket
        }
        return brackets.last
    }
}
